File: src/harness_core/dispatch/routing/dynamic_tier_selector.py
"""Dynamic tier selector: adaptive routing with cold-start fallback."""

from __future__ import annotations

from typing import Any

from ..dispatch_decision import MODEL_TIERS, RejectedCandidate, ShadowRoute
from ..model_selector import ModelSelector
from ..task_analyzer import TaskAnalysis
from .cost_of_pass_router import CostOfPassRouter
from .promotion_gate import PromotionGate
from .schemas import RoutingSelection, make_task_group


class DynamicTierSelector:
    """Select tier dynamically per task group, falling back to static rules."""

    def __init__(
        self,
        static_selector: ModelSelector,
        cost_of_pass_router: CostOfPassRouter,
        promotion_gate: PromotionGate,
    ) -> None:
        self._static = static_selector
        self._router = cost_of_pass_router
        self._gate = promotion_gate

    def select(
        self, analysis: TaskAnalysis
    ) -> RoutingSelection:
        task_group = make_task_group(analysis.task_domain, analysis.task_intent)

        if self._router.can_route_adaptively(task_group):
            result = self._router.best_tier_for_task_group(task_group)
            if result is not None:
                best_tier, _cop = result
                verdict = self._gate.evaluate(task_group, best_tier)
                if verdict.verdict == "promote":
                    rejected: list[RejectedCandidate] = []
                    reasons: list[str] = [f"adaptive_routing:cost_of_pass"]

                    selected_tier, rejected, reasons = self._apply_hard_constraints(
                        analysis, best_tier, rejected, reasons,
                    )

                    fallback_tier = self._fallback_tier(selected_tier)
                    shadow_routes = self._build_shadow_routes(analysis, selected_tier, fallback_tier)

                    return RoutingSelection(
                        selected_tier=selected_tier,
                        selected_profile_id=None,
                        fallback_tier=fallback_tier,
                        fallback_profile_id=None,
                        shadow_routes=shadow_routes,
                        rejected_candidates=rejected,
                        routing_reason="; ".join(reasons),
                        routing_mode="adaptive",
                    )

        # Cold-start fallback: delegate to static selector, wrap in RoutingSelection
        static_result = self._static.select(analysis)
        return RoutingSelection(
            selected_tier=static_result[0],
            selected_profile_id=static_result[1],
            fallback_tier=static_result[2],
            fallback_profile_id=static_result[3],
            shadow_routes=static_result[4],
            rejected_candidates=static_result[5],
            routing_reason=f"adaptive_cold_start_fallback; {static_result[6]}",
            routing_mode="static",
        )

    def _apply_hard_constraints(
        self,
        analysis: TaskAnalysis,
        selected_tier: str,
        rejected: list[RejectedCandidate],
        reasons: list[str],
    ) -> tuple[str, list[RejectedCandidate], list[str]]:
        if analysis.confidence_label == "low":
            reasons.append("low_confidence_escalation")
            rejected.append(RejectedCandidate(
                tier=selected_tier,
                profile_id=None,
                reason="low confidence",
                constraint_failed="confidence_threshold",
            ))
            selected_tier = "strong_planner"

        if analysis.risk_level == "critical":
            if selected_tier not in ("strong_planner", "advisor"):
                rejected.append(RejectedCandidate(
                    tier=selected_tier,
                    profile_id=None,
                    reason="critical risk requires stronger tier",
                    constraint_failed="risk_level",
                ))
                selected_tier = "strong_planner"
                reasons.append("critical_risk_override")

        if analysis.context_budget_estimate < 500:
            rejected.append(RejectedCandidate(
                tier="strong_planner",
                profile_id=None,
                reason="budget too low for strong_planner",
                constraint_failed="budget_threshold",
            ))
            reasons.append("budget_constrained")

        return selected_tier, rejected, reasons

    def _fallback_tier(self, selected: str) -> str:
        tier_order = list(MODEL_TIERS)
        idx = tier_order.index(selected) if selected in tier_order else 1
        if idx < len(tier_order) - 1:
            return tier_order[idx + 1]
        return tier_order[-1]

    def _build_shadow_routes(
        self, analysis: TaskAnalysis, selected: str, fallback: str
    ) -> list[ShadowRoute]:
        routes: list[ShadowRoute] = []
        if fallback != selected:
            routes.append(ShadowRoute(
                tier=fallback,
                profile_id=None,
                reason="fallback option",
                admission_scope="diagnostic",
                expected_tradeoff="lower cost, potentially lower quality",
            ))
        if selected != "cheap_executor":
            routes.append(ShadowRoute(
                tier="cheap_executor",
                profile_id=None,
                reason="cost-optimized alternative",
                admission_scope="diagnostic",
                expected_tradeoff="lowest cost, adequate for simple tasks",
            ))
        if not routes:
            routes.append(ShadowRoute(
                tier=selected,
                profile_id=None,
                reason="self-diagnostic (no cheaper alternative)",
                admission_scope="diagnostic",
                expected_tradeoff="same tier, diagnostic comparison",
            ))
        return routes

File: engine/src/routing/dynamic_tier_selector.rs
use crate::dispatch_decision::{RejectedCandidate, ShadowRoute, MODEL_TIERS};
use crate::model_selector::ModelSelector;
use crate::task_analyzer::TaskAnalysis;

use super::cost_of_pass_router::CostOfPassRouter;
use super::history_store::RoutingHistoryStore;
use super::promotion_gate::{PromotionGate, RoutingObservationStore};
use super::schemas::{make_task_group, RoutingSelection};

pub struct DynamicTierSelector {
    cost_router: CostOfPassRouter,
    promotion_gate: PromotionGate,
}

impl DynamicTierSelector {
    pub fn new(cost_router: CostOfPassRouter, promotion_gate: PromotionGate) -> Self {
        Self {
            cost_router,
            promotion_gate,
        }
    }

    pub fn select(
        &self,
        analysis: &TaskAnalysis,
        history_store: &mut RoutingHistoryStore,
        observation_store: &RoutingObservationStore,
        static_selector: &ModelSelector,
    ) -> RoutingSelection {
        let task_group = make_task_group(&analysis.task_domain, &analysis.task_intent);

        if self
            .cost_router
            .can_route_adaptively(history_store, &task_group)
        {
            if let Some((best_tier, _cop)) = self
                .cost_router
                .best_tier_for_task_group(history_store, &task_group)
            {
                let verdict = self.promotion_gate.evaluate(
                    observation_store,
                    &task_group,
                    &best_tier,
                    "balanced_worker",
                );
                if verdict.verdict == "promote" {
                    let mut rejected: Vec<serde_json::Value> = Vec::new();
                    let reasons: Vec<String> = vec!["adaptive_routing:cost_of_pass".to_string()];

                    let (selected_tier, rejected_typed, reasons_out) =
                        apply_hard_constraints(analysis, &best_tier, rejected, reasons);
                    rejected = rejected_typed;
                    let reasons = reasons_out;

                    let fallback_tier = fallback_tier(&selected_tier);
                    let shadow_routes = build_shadow_routes(&selected_tier, &fallback_tier);

                    return RoutingSelection {
                        selected_tier,
                        selected_profile_id: None,
                        fallback_tier,
                        fallback_profile_id: None,
                        shadow_routes,
                        rejected_candidates: rejected,
                        routing_reason: reasons.join("; "),
                        routing_mode: "adaptive".to_string(),
                        routing_experiment_id: None,
                    };
                }
            }
        }

        // Cold-start fallback: delegate to static selector
        let static_result = static_selector.select(analysis);
        RoutingSelection {
            selected_tier: static_result.selected_tier.clone(),
            selected_profile_id: static_result.selected_profile_id.clone(),
            fallback_tier: static_result.fallback_tier.clone(),
            fallback_profile_id: static_result.fallback_profile_id.clone(),
            shadow_routes: static_result
                .shadow_routes
                .iter()
                .map(|r| serde_json::to_value(r).unwrap_or_default())
                .collect(),
            rejected_candidates: static_result
                .rejected_candidates
                .iter()
                .map(|r| serde_json::to_value(r).unwrap_or_default())
                .collect(),
            routing_reason: format!(
                "adaptive_cold_start_fallback; {}",
                static_result.routing_reason
            ),
            routing_mode: "static".to_string(),
            routing_experiment_id: None,
        }
    }
}

fn apply_hard_constraints(
    analysis: &TaskAnalysis,
    selected_tier: &str,
    mut rejected: Vec<serde_json::Value>,
    mut reasons: Vec<String>,
) -> (String, Vec<serde_json::Value>, Vec<String>) {
    let mut tier = selected_tier.to_string();

    if analysis.confidence_label == "low" {
        reasons.push("low_confidence_escalation".to_string());
        rejected.push(
            serde_json::to_value(&RejectedCandidate {
                tier: tier.clone(),
                profile_id: None,
                reason: "low confidence".to_string(),
                constraint_failed: Some("confidence_threshold".to_string()),
                estimated_cost: None,
            })
            .unwrap_or_default(),
        );
        tier = "strong_planner".to_string();
    }

    if analysis.risk_level == "critical" && tier != "strong_planner" && tier != "advisor" {
        rejected.push(
            serde_json::to_value(&RejectedCandidate {
                tier: tier.clone(),
                profile_id: None,
                reason: "critical risk requires stronger tier".to_string(),
                constraint_failed: Some("risk_level".to_string()),
                estimated_cost: None,
            })
            .unwrap_or_default(),
        );
        tier = "strong_planner".to_string();
        reasons.push("critical_risk_override".to_string());
    }

    if analysis.context_budget_estimate < 500 {
        rejected.push(
            serde_json::to_value(&RejectedCandidate {
                tier: "strong_planner".to_string(),
                profile_id: None,
                reason: "budget too low for strong_planner".to_string(),
                constraint_failed: Some("budget_threshold".to_string()),
                estimated_cost: None,
            })
            .unwrap_or_default(),
        );
        reasons.push("budget_constrained".to_string());
    }

    (tier, rejected, reasons)
}

fn fallback_tier(selected: &str) -> String {
    let tier_order: Vec<&str> = MODEL_TIERS.to_vec();
    let idx = tier_order.iter().position(|t| *t == selected).unwrap_or(1);
    if idx < tier_order.len() - 1 {
        tier_order[idx + 1].to_string()
    } else {
        tier_order.last().unwrap().to_string()
    }
}

fn build_shadow_routes(selected: &str, fallback: &str) -> Vec<serde_json::Value> {
    let mut routes: Vec<serde_json::Value> = Vec::new();
    if fallback != selected {
        routes.push(
            serde_json::to_value(&ShadowRoute {
                tier: fallback.to_string(),
                profile_id: None,
                reason: "fallback option".to_string(),
                admission_scope: "diagnostic".to_string(),
                estimated_cost: None,
                expected_tradeoff: "lower cost, potentially lower quality".to_string(),
            })
            .unwrap_or_default(),
        );
    }
    if selected != "cheap_executor" {
        routes.push(
            serde_json::to_value(&ShadowRoute {
                tier: "cheap_executor".to_string(),
                profile_id: None,
                reason: "cost-optimized alternative".to_string(),
                admission_scope: "diagnostic".to_string(),
                estimated_cost: None,
                expected_tradeoff: "lowest cost, adequate for simple tasks".to_string(),
            })
            .unwrap_or_default(),
        );
    }
    if routes.is_empty() {
        routes.push(
            serde_json::to_value(&ShadowRoute {
                tier: selected.to_string(),
                profile_id: None,
                reason: "self-diagnostic (no cheaper alternative)".to_string(),
                admission_scope: "diagnostic".to_string(),
                estimated_cost: None,
                expected_tradeoff: "same tier, diagnostic comparison".to_string(),
            })
            .unwrap_or_default(),
        );
    }
    routes
}
